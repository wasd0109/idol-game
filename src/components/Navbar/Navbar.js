import React from 'react';
import { Link, useLocation } from 'react-router-dom';

const generateNavBarButton = (name, link, currentLocation) => (
  <Link
    key={name}
    to={link}
    className={`transition duration-500 hover:bg-white h-full pt-4 pr-2 ${
      currentLocation === link ? 'bg-white' : ''
    }`}
    id={link}
  >
    <p className="text-lg font-medium">{name}</p>
  </Link>
);

function Navbar({ onLogout, navBarContent }) {
  const location = useLocation();
  return (
    <nav className="flex flex-wrap bg-blue-300 pl-4" id="navbar">
      {navBarContent.map((button) =>
        generateNavBarButton(button[0], button[1], location.pathname)
      )}
      <Link
        to="/"
        className="transition duration-500 hover:bg-white h-full pt-4 px-2 mr-1 ml-auto"
        onClick={onLogout}
      >
        <p className="text-lg font-medium">Logout</p>
      </Link>
    </nav>
  );
}

export default Navbar;
