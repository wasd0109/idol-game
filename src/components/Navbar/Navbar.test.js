import React from 'react';
import { render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import Navbar from './Navbar';

const initialProps = {
    onLogout: jest.fn(),
};

const navBarContent = [
    ['Home', '/'],
    ['Player List', '/players'],
    ['Battle', '/battle'],
    ['Setting', '/setting'],
    ["Logout", "/"]
]

let component;

beforeEach(() => {
    component = render(<Router><Navbar {...initialProps} /></Router>);
});

describe('Navbar component render correctly', () => {
    test('Navbar component render', () => {

        expect(component).toMatchSnapshot();
    });


    test.each(navBarContent)("Navbar render correct link", (linkName, linkRef) => {
        expect(screen.getByText(linkName)).toBeInTheDocument();
        expect(screen.getByText(linkName).parentElement.getAttribute("href")).toEqual(linkRef);
    })
});

describe("Navbar component function correctly", () => {
    test.each(navBarContent)("Navbar link redirect to correct path", (linkName, linkRef) => {
        const url = `http://localhost${linkRef}`;
        userEvent.click(screen.getByText(linkName));
        expect(document.URL).toEqual(url);
    })
    test("Clicking logout button trigger logout", () => {
        userEvent.click(screen.getByText("Logout"));
        expect(initialProps.onLogout).toBeCalled();
    })
})
