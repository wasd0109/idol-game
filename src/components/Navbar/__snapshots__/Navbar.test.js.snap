// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar component render correctly Navbar component render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="flex flex-wrap bg-blue-300 pl-4"
        id="navbar"
      >
        <a
          class="transition duration-500 hover:bg-white h-full pt-4 pr-2"
          href="/"
        >
          <p
            class="text-lg font-medium"
          >
            Home
          </p>
        </a>
        <a
          class="transition duration-500 hover:bg-white h-full pt-4 pr-2"
          href="/players"
        >
          <p
            class="text-lg font-medium"
          >
            Player List
          </p>
        </a>
        <a
          class="transition duration-500 hover:bg-white h-full pt-4 pr-2"
          href="/battle"
        >
          <p
            class="text-lg font-medium"
          >
            Battle
          </p>
        </a>
        <a
          class="transition duration-500 hover:bg-white h-full pt-4 pr-2"
          href="/setting"
        >
          <p
            class="text-lg font-medium"
          >
            Setting
          </p>
        </a>
        <a
          class="transition duration-500 hover:bg-white h-full pt-4 px-2 mr-1 ml-auto"
          href="/"
        >
          <p
            class="text-lg font-medium"
          >
            Logout
          </p>
        </a>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="flex flex-wrap bg-blue-300 pl-4"
      id="navbar"
    >
      <a
        class="transition duration-500 hover:bg-white h-full pt-4 pr-2"
        href="/"
      >
        <p
          class="text-lg font-medium"
        >
          Home
        </p>
      </a>
      <a
        class="transition duration-500 hover:bg-white h-full pt-4 pr-2"
        href="/players"
      >
        <p
          class="text-lg font-medium"
        >
          Player List
        </p>
      </a>
      <a
        class="transition duration-500 hover:bg-white h-full pt-4 pr-2"
        href="/battle"
      >
        <p
          class="text-lg font-medium"
        >
          Battle
        </p>
      </a>
      <a
        class="transition duration-500 hover:bg-white h-full pt-4 pr-2"
        href="/setting"
      >
        <p
          class="text-lg font-medium"
        >
          Setting
        </p>
      </a>
      <a
        class="transition duration-500 hover:bg-white h-full pt-4 px-2 mr-1 ml-auto"
        href="/"
      >
        <p
          class="text-lg font-medium"
        >
          Logout
        </p>
      </a>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
